x-pack/plugins/security/server/lib/authorization/has_privileges.test.js

```diff
@@ -0,0 +1,311 @@
+  const request = Symbol();
+  const hasPrivileges = hasPrivilegesWithRequest(request);
+  const privileges = [privilege];
+  const result = await hasPrivileges(privileges);
+
+  expectNoDeprecationLogged(mockServer);
+  expect(mockCallWithRequest).toHaveBeenCalledWith(request, 'shield.hasPrivileges', {
+    body: {
+      applications: [{
+        application: defaultApplication,
+        resources: [DEFAULT_RESOURCE],
+        privileges: [
+          getVersionPrivilege(defaultVersion), getLoginPrivilege(), ...privileges
+        ]
+      }]
+    }
+  });
+  expect(result).toEqual({
+    result: HAS_PRIVILEGES_RESULT.AUTHORIZED,
+    username,
+    missing: [],
+  });
+});
+
+test(`returns unauthorized they have only one application privilege`, async () => {
+  const privilege1 = `action:saved_objects/${savedObjectTypes[0]}/get`;
+  const privilege2 = `action:saved_objects/${savedObjectTypes[0]}/create`;
+  const username = 'foo-username';
+  const mockServer = createMockServer();
+  const mockCallWithRequest = createMockCallWithRequest([
+    mockApplicationPrivilegeResponse({
+      hasAllRequested: false,
+      privileges: {
+        [getVersionPrivilege(defaultVersion)]: true,
+        [getLoginPrivilege()]: true,
+        [privilege1]: true,
+        [privilege2]: false,
+      },
+      application: defaultApplication,
+      username,
+    })
+  ]);
+
+  const hasPrivilegesWithRequest = hasPrivilegesWithServer(mockServer);
+  const request = Symbol();
+  const hasPrivileges = hasPrivilegesWithRequest(request);
+  const privileges = [privilege1, privilege2];
+  const result = await hasPrivileges(privileges);
+
+  expectNoDeprecationLogged(mockServer);
+  expect(mockCallWithRequest).toHaveBeenCalledWith(request, 'shield.hasPrivileges', {
+    body: {
+      applications: [{
+        application: defaultApplication,
+        resources: [DEFAULT_RESOURCE],
+        privileges: [
+          getVersionPrivilege(defaultVersion), getLoginPrivilege(), ...privileges
+        ]
+      }]
+    }
+  });
+  expect(result).toEqual({
+    result: HAS_PRIVILEGES_RESULT.UNAUTHORIZED,
+    username,
+    missing: [privilege2],
+  });
+});
+
+test(`throws error if missing version privilege and has login privilege`, async () => {
+  const privilege = `action:saved_objects/${savedObjectTypes[0]}/get`;
+  const mockServer = createMockServer();
+  createMockCallWithRequest([
+    mockApplicationPrivilegeResponse({
+      hasAllRequested: false,
+      privileges: {
+        [getVersionPrivilege(defaultVersion)]: false,
+        [getLoginPrivilege()]: true,
+        [privilege]: true,
+      }
+    })
+  ]);
+
+  const hasPrivilegesWithRequest = hasPrivilegesWithServer(mockServer);
+  const hasPrivileges = hasPrivilegesWithRequest({});
+
+  await expect(hasPrivileges([privilege])).rejects.toThrowErrorMatchingSnapshot();
+  expectNoDeprecationLogged(mockServer);
+});
+
+describe('legacy fallback with no application privileges', () => {
```

Comment:
I think we need a set of tests for `legacy fallback with application privileges` as well. This ensures that we don't accidentally introduce a bug where legacy privileges trump the new model. They'd probably be the same tests as above, only they'd also include the legacy privileges. And we wouldn't want to just add the legacy privileges to the app privilege tests because then we wouldn't have any test in place for only app privileges, which is where all new users would be.

Reply:
The way the tests are currently written, if we tried to retrieve the legacy index privileges whenever the application privileges were already returned and sufficient, we'd hit an unprogrammed mock and an error would be thrown (as we aren't mocking the response for the has_privileges index call).

I could add dedicated "index privileges and application privileges" tests, but we'd be stubbing out `callWithRequest` calls that would never be made.